### PR TITLE
Fix issue #2319: Open Java extension log doesn't open latest file

### PR DESCRIPTION
**Pull request by quanzhuo**

### CODE REVIEW

**Comment by quanzhuo** [comment]
Fixes #2319

**Comment by quanzhuo** [comment]
@rgrunber Could you help to review this ?

**Review by rgrunber — CHANGES_REQUESTED** [review]
It definitely works, but would you be able to rewrite this in terms of overriding the `sort` function ? eg.

```
files.sort((a, b) => {
...
sorting logic here
...
}
```

I just think it's a bit simpler to read because you only need to compare 2 values (the date if not the same, and the extension number for identical dates). If it becomes too annoying I'm happy to just review this approach also.

**Comment by quanzhuo** [comment]
> It definitely works, but would you be able to rewrite this in terms of overriding the `sort` function ? eg.
> 
> ```
> files.sort((a, b) => {
> ...
> sorting logic here
> ...
> }
> ```
> 
> I just think it's a bit simpler to read because you only need to compare 2 values (the date if not the same, and the extension number for identical dates). If it becomes too annoying I'm happy to just review this approach also.

Thanks for your advise. I have changed the logic, it looks better now. @rgrunber

File: src/extension.ts [hunk 1]
```diff
@@ -727,7 +727,21 @@ function openClientLogFile(logFile: string, column: ViewColumn = ViewColumn.Acti
 		// find out the newest one
 		glob(filename + '.*', { cwd: dirname }, (err, files) => {
 			if (!err && files.length > 0) {
-				files.sort();
+				files.sort((a, b) => {
+					let date1 = a.slice(11, 21), date2 = b.slice(11, 21);
+					if (date1 < date2) {
```
[the remark this answers is not in the document]
Comment (quanzhuo) [reply src/extension.ts:732]: > How about
> 
> ```js
> if (data1 === date2) {
>     ...
> } else {
>     return date1 - date2;
> }
> ```
> 
> You can directly return `a - b` or `b - a` in sort comparator to make the code simple.

Thank for advise, but I am not quite clear about your idea. `date1` and `date2` are strings like `2022-01-18` .  Can I apply `-` between strings? @jdneo

[the remark this answers is not in the document]
Comment (jdneo) [reply src/extension.ts:732]: You could use `new Date(a) - new Date(b)`. But that will lose some performance: https://jsben.ch/Ip2Nf.

At least line 737 and line 739 could be updated.

[the remark this answers is not in the document]
Comment (quanzhuo) [reply src/extension.ts:732]: line 737 and 739 updated @jdneo

**Comment by rgrunber** [comment]
Change looks good to me. Just 2 minor things  :
- Instead of `date1`, `date2`, `ext1`, `ext2`, I would go with `dateA`, `dateB`, `extA`, `extB` to connect those values to their respective parameters (`a` & `b`).
- You could combine `date1 < date2` and the `else` case into 1 statement by restructuring as follows :
```
if (date1 === date2) {
  ... stays the same ...
} else {
  return date1 < date2 ? -1 : 1;
}
```

I'll leave it to @jdneo to confirm he has no other comments, but once the above is addressed, I think we can merge.

**Comment by quanzhuo** [comment]
All have updated as you suggested.  Your suggestions really helps. @rgrunber @jdneo

**Review by jdneo — APPROVED** [review]
LGTM 👍

**Review by rgrunber — APPROVED** [review]

**Pull request merged**